### ISSUE TITLE
New feature: `$find`

### ISSUE THREAD

**Issue by noahlaux** [json-e/json-e#525]
## $find

This issue explores the possibility of adding a `$find`  operator as found in many functional languages, as a sibling to the [`$map`](https://json-e.js.org/operators.html#map) operator following the same interface.

While it's possible to achieve the same using combinations of other operators currently available, it comes with the price of quite some boilerplate which can be confusing in bigger logical blocks (though granted, more flexible). This becomes clear if we want to find a single entry in a list, especially combined with `$let`.

### Here's a barebone example of how to achieve this currently in `json-e`:

```yaml
$let:
  haystack: 
    $map: [1, 2, 3]
    each(entry):
      $if: entry == 2
      then:
        $eval: entry
  in:
    # I know technically this last $let section isn't strictly necessary, but imagine fx 
    # if we want to refer `needle` in multiple places instead of haystack[0] 
    $let:
      needle: haystack[0]
    in:
      $eval: needle 
      # => 2
```

### Here's an example of how the same could be achieved with `$find`

```yaml
$let:
  needle:
    $find: [1, 2, 3]
    each(entry): entry == 2
in:
  $eval: needle
  # => 2
```

What do y'all think? Just syntax sugar or actually useful?
If you like it I can write up some specs and do some implementations.

**Comment by djmitche** [comment]
I like this idea! It would differ a little from `$map` in that it is only over arrays, since objects are un-ordered.

We would need to define what this renders to when no entry matches. Is that an error, or `null`, or deletion marker?

I can see this being extended with a `reverse: true` to scan the array in reverse, and with a `default: ..` for when no entry matches. That could come later, of course.

**Comment by noahlaux** [comment]
> I like this idea! It would differ a little from $map in that it is only over arrays, since objects are un-ordered.

@djmitche Glad you liked it! And I thought the same about objects!

> We would need to define what this renders to when no entry matches. Is that an error, or null, or deletion marker?

So the textbook implementation for a `find` should return the first element in the provided array that satisfies the provided testing function. If no values satisfy the testing function, `null` is returned. Of course we should think about how to adapt this to a `json-e` context.

I wouldn't expect it to error if no candidate satisfies the testing function. At least for me, as I use `json-e` in a dynamic environment where this is a feasible scenario. I'd rather like to have consumers guard themselves, fx by using [`defined`](https://json-e.js.org/built-ins.html#context). I think this could already happen with people using fx `$let` where a variable could potentially be `null`.

I like the idea of using the deletion marker, if this is what you mean:

### Find with a match
```yaml
# Template
{ a: 1, b: { $find: haystack, each(x): x == 3 }}}
# Context
{ haystack: [1, 2, 3] }
# result
{ a: 1, b: 3}
```

### Find without a match result in deletion marker

```yaml
# Template
{ a: 1, b: { $find: haystack, each(x): x > 4 }}}
# Context
{ haystack: [1, 2, 3] }
# result
{ a: 1 }
```

I'm slightly unsure whether `each` is not the best word describing the testing function, only it's slightly convinient it matches that of `$map`. Alternatives could be `fn`, `cb`, `test`, `cond` (for condition).

> I can see this being extended with a reverse: true to scan the array in reverse, and with a default: .. for when no entry matches. That could come later, of course.

Hmm, I'm thinking it could just reverse the provided array with [$reverse](https://json-e.js.org/operators.html#reverse)

**Comment by djmitche** [comment]
I like the deletion marker better than `null` -- and yes, that's precisely what I mean. We've tried (not entirely successfully) not to treat `null` as special, but as just another JSON value.

I think `each` is fine, and using an alternative would just be one more thing for users to remember.

Good point about `$reverse`!

**Comment by noahlaux** [comment]
@djmitche I tried to compile some specifications, maybe you have some more suggestions?

```yaml
---
section: $find operator
---
title:   $find, simple find
context: {}
template: {$find: [2, 4, 6], each(x): 'x == 2' }
result: 2
---
title:   $find, simple find with context
context: {a: 2}
template: {$find: [2, 4, 6], each(x): 'x == a' }
result: 2
---
title:   $find respects delete marker
context: {}
template: {a: 1, b: {$find: [1,2,3], 'each(x)': 'x > 4' }}
result: {a: 1}
---
title:   $find, complicated find with $let
context: {}
template:
  $let:
    needle:
      $find: [1, 2, 3]
      each(entry): entry == 2
  in:
    $eval: needle
result: 2
---
title:    $find complex identifier
context:  {a: 1}
template:
  $find: [2, 4, 6]
  each(my_identifier97): 'my_identifier97 == 1 || my_identifier97 == 4'
result: 4
---
title:    $find array index variable
context: {}
template:
  $find: [1, 2, 3]
  each(y,i): 'i == 2'
result: 3
---
title:    $find no args
context:  {}
template:
  $find: [1, 2, 3]
  each(): 'y.k'
error: 'TemplateError: $find has undefined properties: each()'
---
title:    $find too many args
context:  {}
template:
  $find: [1, 2, 3]
  each(k,v,i): 'y.k'
error: 'TemplateError: $find has undefined properties: each(k,v,i)'
---
title:    $find requires an array, not string
context:  {}
template:
  $find: "a, b, c"
  each(y): 'y.k'
error: 'TemplateError: $find value must evaluate to an array'
---
title:    $map requires an array, not number
context:  {}
template:
  $find: 10.1
  each(y): 'y.k'
error: 'TemplateError: $find value must evaluate to an array'
---
title:    $map requires an array, not null
context:  {}
template:
  $find: null
  each(y): 'y.k'
error: 'TemplateError: $find value must evaluate to an array'
---
title:    $find with undefined properties
context:  {a: 1}
template:
  $find: [2, 4, 6]
  each(x): 'x + a'
  foo: 'bar'
error: 'TemplateError: $find has undefined properties: foo'
---
title:    $find each must be string, not number
context:  {}
template:
  $find: [1,2,3]
  each(a): 1
error: 'TemplateError: $each can evaluate string expressions only'
---
title:    $find each must be string, not object
context:  {}
template:
  $find: [1,2,3]
  each(a): {a: 1}
error: 'TemplateError: $each can evaluate string expressions only'
```

**Comment by djmitche** [comment]
That's a good start! How about `$map requires an array, not object` too :)

**Comment by noahlaux** [comment]
@djmitche I suppose once releasing this we can close this issue

**Comment by djmitche** [comment]
released in 4.7.0
